File: src/main/java/io/github/shaksternano/borgar/util/tenor/TenorUtil.java
package io.github.shaksternano.borgar.util.tenor;

import com.google.gson.JsonParser;

import java.net.URI;
import java.net.URISyntaxException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.concurrent.CompletableFuture;

/**
 * For dealing with the Tenor API.
 */
public class TenorUtil {

    /**
     * Retrieves the direct file URL from a Tenor link.
     *
     * @param url       The Tenor URL.
     * @param mediaType The media type to get the URL of.
     * @param apiKey    The Tenor API key to use.
     * @return A future containing the direct file URL.
     */
    public static CompletableFuture<String> retrieveTenorMediaUrl(String url, String mediaType, String apiKey) {
        var invalidUrlMessage = "Invalid Tenor URL";
        try {
            var uri = new URI(url);
            if (uri.getHost().contains("tenor.com") && uri.getPath().startsWith("/view/")) {
                var mediaId = url.substring(url.lastIndexOf("-") + 1);
                var requestUrl = "https://g.tenor.com/v1/gifs?key=" + apiKey + "&ids=" + mediaId;
                var request = HttpRequest.newBuilder()
                    .uri(new URI(requestUrl))
                    .build();
                return HttpClient.newHttpClient()
                    .sendAsync(request, HttpResponse.BodyHandlers.ofString())
                    .thenApply(response -> JsonParser.parseString(response.body())
                        .getAsJsonObject()
                        .getAsJsonArray("results")
                        .get(0)
                        .getAsJsonObject()
                        .getAsJsonArray("media")
                        .get(0)
                        .getAsJsonObject()
                        .getAsJsonObject(mediaType)
                        .getAsJsonPrimitive("url")
                        .getAsString()
                    );
            } else {
                throw new IllegalArgumentException(invalidUrlMessage);
            }
        } catch (URISyntaxException e) {
            throw new IllegalArgumentException(invalidUrlMessage, e);
        }
    }

    public static CompletableFuture<String> retrieveTenorMediaUrl(String url, TenorMediaType mediaType, String apiKey) {
        return retrieveTenorMediaUrl(url, mediaType.getKey(), apiKey);
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/util/TenorUtil.kt
package io.github.shaksternano.borgar.core.util

import io.github.shaksternano.borgar.core.io.UrlInfo
import io.github.shaksternano.borgar.core.io.httpGet
import kotlinx.serialization.Serializable
import java.net.URI

private const val DEFAULT_TENOR_API_KEY: String = "LIVDSRZULELA"

suspend fun retrieveTenorMediaUrl(url: String, getGif: Boolean): UrlInfo? {
    val mediaType =
        if (getGif) TenorMediaType.GIF_LARGE
        else TenorMediaType.MP4_NORMAL
    val mediaUrl = retrieveTenorMediaUrl(url, mediaType) ?: return null
    return UrlInfo(
        url = mediaUrl,
        gifv = !getGif,
    )
}

suspend fun retrieveTenorMediaUrl(url: String, mediaType: TenorMediaType): String? {
    if (!isTenorUrl(url)) return null
    val apiKey = getEnvVar("TENOR_API_KEY") ?: DEFAULT_TENOR_API_KEY
    val mediaId = url.substring(url.lastIndexOf("-") + 1)
    val requestUrl = "https://g.tenor.com/v1/gifs?key=$apiKey&ids=$mediaId"
    val responseBody = httpGet<TenorResponse>(requestUrl)
    return runCatching {
        responseBody.results.first().media.first().getOrElse(mediaType.id) {
            throw IllegalArgumentException("Media type $mediaType not found")
        }.url
    }.getOrElse {
        throw IllegalStateException("Error parsing Tenor response", it)
    }
}

fun isTenorUrl(url: String): Boolean {
    val uri = runCatching { URI(url) }.getOrElse { return false }
    val host = uri.host
    return host != null && host.contains("tenor.com") && uri.getPath().startsWith("/view/")
}

/**
 * Contains Tenor string constants for different media types.
 */
@Suppress("unused")
enum class TenorMediaType(
    override val id: String,
    override val displayName: String
) : Identified, Displayed {
    GIF_EXTRA_SMALL("nanogif", "Extra small GIF"),
    GIF_SMALL("tinygif", "Small GIF"),
    GIF_NORMAL("gif", "Normal GIF"),
    GIF_LARGE("mediumgif", "Large GIF"),

    MP4_EXTRA_SMALL("nanomp4", "Extra small MP4"),
    MP4_SMALL("tinymp4", "Small MP4"),
    MP4_NORMAL("mp4", "Normal MP4"),
    MP4_NORMAL_LOOPED("loopedmp4", "Normal looped MP4"),

    WEBM_EXTRA_SMALL("nanowebm", "Extra small WebM"),
    WEBM_SMALL("tinywebm", "Small WebM"),
    WEBM_NORMAL("webm", "Normal WebM"),
}

@Serializable
private data class TenorResponse(
    val results: List<Result>,
)

@Serializable
private data class Result(
    val media: List<Map<String, Media>>,
)

@Serializable
private data class Media(
    val url: String,
)
